lib/msf/core/exploit/remote/http/nagios_xi/login.rb

```diff
@@ -0,0 +1,89 @@
+# -*- coding: binary -*-
+
+module Msf::Exploit::Remote::HTTP::NagiosXi::Login
+  # performs a Naxios XI login
+  #
+  # @param user [String] Username
+  # @param pass [String] Password
+  # @return [String, Msf::Exploit::CheckCode] String containing the session cookies on successful login, Msf::Exploit::CheckCode otherwise
+  def nagios_xi_login(user, pass)
+    # empty the cookie jar, otherwise the module will start failing if it is run multiple times in a short timespan
+    self.cookie_jar = [].to_set
+
+    # visit the login page in order to obtain the cookies and the `nsp_str` token required for authentication
+    res_pre_login = send_request_cgi({
+      'method' => 'GET',
+      'uri' => nagios_xi_login_url,
+      'keep_cookies' => true,
+    })
+
+    unless res_pre_login
+      return Msf::Exploit::CheckCode::Unknown('Connection failed')
+    end
+
+    unless res_pre_login.code == 200 && res_pre_login.body.include?('<b>Nagios XI</b>')
+      return Msf::Exploit::CheckCode::Safe('Target is not a Nagios XI application')
+    end
```

Comment:
This check is insufficient.

For example, on Nagios XI 5.2.3:

```
# curl -isk https://10.1.1.109/nagiosxi/login.php | grep Nagios
                <!-- Produced by Nagios XI. Copyright (c) 2008-2021 Nagios Enterprises, LLC (www.nagios.com). All Rights Reserved. -->
        <!-- Powered by the Nagios Synthesis Framework -->
                <title>Nagios XI - Login</title>
            <img src="https://10.1.1.109/nagiosxi/images/nagios_logo_white_transbg.png" border="0" class="xi-logo" alt="Nagios XI" title="Nagios XI">
        <h3>About Nagios XI</h3>
            Nagios XI is an enterprise-class monitoring and alerting solution that provides organizations with extended insight of their IT infrastructure before problems affect critical business processes.  For more information on Nagios XI, visit the <a href="//www.nagios.com/products/nagiosxi/" target="_blank">Nagios XI product page</a>.
        <h3>Nagios Learning Opportunities</h3>
            Learn about Nagios            <a href="//www.nagios.com/services/training"
            Want to learn about how other experts are utilizing Nagios?  Don't miss your chance to attend the next            <a href="//go.nagios.com/nwcna" target="_blank"><strong>Nagios World Conference</strong></a>.
                <a href="http://nagios.com/products/nagiosxi" target="new"><strong>Nagios XI</strong></a>                             </div>
                Copyright &copy; 2008-2021 <a href="https://www.nagios.com/" target="_blank">Nagios Enterprises, LLC</a>
        <h2>Nagios XI</h2>
            Copyright (c) 2009-2021 Nagios Enterprises, LLC<br>
```

```
# curl -isk https://10.1.1.109/nagiosxi/login.php | grep '>Nagios XI<'
                <a href="http://nagios.com/products/nagiosxi" target="new"><strong>Nagios XI</strong></a>                             </div>
        <h2>Nagios XI</h2>
```

<details><p>

```
msf6 auxiliary(scanner/http/nagios_xi_scanner) > set httptrace true
httptrace => true
msf6 auxiliary(scanner/http/nagios_xi_scanner) > run

####################
# Request:
####################
GET /nagiosxi/login.php HTTP/1.1
Host: 10.1.1.109
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)


####################
# Response:
####################
HTTP/1.0 200 OK
Date: Sat, 06 Feb 2021 02:19:00 GMT
Server: Apache/2.2.15 (CentOS)
X-Powered-By: PHP/5.3.3
Set-Cookie: nagiosxi=eeuss0ar2amete9oinubvdelv6; expires=Sat, 06-Feb-2021 02:49:00 GMT; path=/
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Connection: close
Content-Type: text/html; charset=UTF-8

        <!DOCTYPE html>
        <!-- <!DOCTYPE html> -->
        <html>

    <head>
        <meta http-equiv="X-UA-Compatible" content="IE=Edge"/>
                <!-- Produced by Nagios XI. Copyright (c) 2008-2021 Nagios Enterprises, LLC (www.nagios.com). All Rights Reserved. -->
        <!-- Powered by the Nagios Synthesis Framework -->
                <title>Nagios XI - Login</title>
        <meta name="ROBOTS" content="NOINDEX, NOFOLLOW">
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>

        
    <link rel="icon" type="image/png" href="https://10.1.1.109/nagiosxi/images/favicon-32x32.png" sizes="32x32" />
    <link rel="shortcut icon" href="https://10.1.1.109/nagiosxi/images/favicon.ico" type="image/ico"/>
    <link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/css/jquery.autocomplete.css'/>

    <script type='text/javascript'>
        var base_url = "https://10.1.1.109/nagiosxi/";
        var backend_url = "https%3A%2F%2F10.1.1.109%2Fnagiosxi%2Flogin.php";
        var ajax_helper_url = "https://10.1.1.109/nagiosxi/ajaxhelper.php";
        var ajax_proxy_url = "https://10.1.1.109/nagiosxi/ajaxproxy.php";
        var suggest_url = "https://10.1.1.109/nagiosxi/suggest.php";
        var request_uri = "%2Fnagiosxi%2Flogin.php";
        var permalink_base = "https://10.1.1.109/nagiosxi/login.php?";
        var demo_mode =0;
        var nsp_str = "56a8f02b80cdf0ca30d8e8d321345533";
    </script>
    
    <!-- Adding Font-Awesome for all themes -->
    <link rel="stylesheet" href="https://10.1.1.109/nagiosxi/includes/css/font-awesome.min.css?1449592399" type="text/css" />

    <!-- main jquery libraries -->
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/jquery/jquery-1.11.2.min.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/jquery/jquery-migrate-1.2.1.min.js?1449592399'></script>
    <link type="text/css" href="https://10.1.1.109/nagiosxi/includes/js/jquery/css/smoothness/jquery-ui-1.9.0.custom.min.css?1449592399" rel="stylesheet"/>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/jquery/jquery.colorBlend.js'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/jquery/jquery.timers-1.1.3.js'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/jquery/jquery-ui-1.9.0.custom.min.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/jquery/jquery-ui-timepicker-addon.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/jquery/jquery.searchabledropdown.custom.js?1449592399'></script>

            <!-- Bootstrap 3 for XI 5 UI -->
        <link rel="stylesheet" href="https://10.1.1.109/nagiosxi/includes/css/bootstrap.3.min.css?1449592399" type="text/css" />
        <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/bootstrap.3.min.js?1449592399'></script>
    
    <!-- spin kit -->
    <link rel="stylesheet" href="https://10.1.1.109/nagiosxi/includes/css/spinkit.css?1449592399" type="text/css" />

    <!-- jquery autocomplete -->
    <link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/css/jquery.autocomplete.css'/>

    <!-- colorpicker -->
    <link rel="stylesheet" href="https://10.1.1.109/nagiosxi/includes/js/jquery/colorpicker/css/colorpicker.css" type="text/css" />
    <script type="text/javascript" src="https://10.1.1.109/nagiosxi/includes/js/jquery/colorpicker/js/colorpicker.js"></script>

    <!-- clipboard plugin -->
    <script type="text/javascript" src="https://10.1.1.109/nagiosxi/includes/js/jquery/jquery.zclip.min.js"></script>

    <!-- XI JS Scripts -->
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/core.js?1449592399'></script>
        <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/themes/modern.js?1449592399'></script>
        
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/commands.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/views.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/dashboards.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/dashlets.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/tables.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/users.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/perfdata.js?1449592399'></script>
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/wizards.js?1449592399'></script>

    <!-- XI CSS -->
    <link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/css/base.css?1449592399' />
        <link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/css/themes/modern.css?1449592399' />
    
    <!-- Highcharts Graphing Library -->
    <script type="text/javascript" src="https://10.1.1.109/nagiosxi//includes/js/highcharts/highcharts.js?1449592399"></script>
            <script type="text/javascript" src="https://10.1.1.109/nagiosxi//includes/js/highcharts/highcharts-more.js?1449592399"></script>
            <script type="text/javascript" src="https://10.1.1.109/nagiosxi//includes/js/highcharts/modules/exporting.js?1449592399"></script>
            <script type="text/javascript" src="https://10.1.1.109/nagiosxi//includes/js/highcharts/modules/no-data-to-display.js?1449592399"></script>
    <!-- D3 Graphing Library -->
    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/js/d3/d3.v3.min.js?1449592399'></script>

    
        <!-- jScrollPane -->
        <link type="text/css" href="https://10.1.1.109/nagiosxi/includes/js/jquery/css/jquery.jscrollpane.css" rel="stylesheet" media="all" />
        <script type="text/javascript" src="https://10.1.1.109/nagiosxi/includes/js/jquery/jquery.jscrollpane.min.js"></script>

    <script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/dashlets/gauges/js/gauge.js'></script>
<link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/dashlets/internethealthreport/internethealthreport.css' />
<link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/dashlets/internettrafficreport/internettrafficreport.css' />
<link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/dashlets/rss_dashlet/rss_dashlet.css' />
<link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/dashlets/sansrisingports/sansrisingports.css' />
<link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/dashlets/worldtimeserver/worldtimeserver.css' />
<script type="text/javascript" src="https://10.1.1.109/nagiosxi/includes/components/graphexplorer/includes/graphexplorerinclude.js"></script><link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/components/helpsystem/css/helpsystem.css?1.1.0' /><link rel='stylesheet' type='text/css' href='https://10.1.1.109/nagiosxi/includes/components/helpsystem/css/bootstrap-tour.min.css?1.1.0' /><script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/components/helpsystem/js/helpsysteminclude.js?1.1.0'></script><script type='text/javascript' src='https://10.1.1.109/nagiosxi/includes/components/helpsystem/js/bootstrap-tour.min.js?1.1.0'></script>    </head>

    <body  class=' parent' >

    
    <div  class="parentpage">

    <div id="whiteout"></div>
    <div id="header" class="parenthead" >
        
<!--- HEADER START -->


    <div id="toplogo">
        <a href="https://10.1.1.109/nagiosxi/" target="_top">
            <img src="https://10.1.1.109/nagiosxi/images/nagios_logo_white_transbg.png" border="0" class="xi-logo" alt="Nagios XI" title="Nagios XI">
            XI
        </a>
    </div>
    
    
    <div id="topmenu">
                <div class="mainmenu">
            <div><a href="https://10.1.1.109/nagiosxi/login.php">Login</a></div>
        </div>
            </div>

    

    <div id="feedback_layer">
        <div id="feedback_content">

            <div id="feedback_close">
                <a id="close_feedback_link" style="display: inline-block;" title="Close"><i class="fa fa-times" style="font-size: 16px;"></i></a>
            </div>

            <div id="feedback_container">

                <div id="feedback_header">
                    <b>Send Us Feedback</b>
                    <p>We love input!  Tell us what you think about this product and you'll directly drive future innovation!</p>
                </div>
                <!-- feedback_header -->

                <div id="feedback_data">

                    <form id="feedback_form" method="get" action="https://10.1.1.109/nagiosxi/ajaxproxy.php">

                        <input type="hidden" name="proxyurl" value="https://api.nagios.com/feedback/">
                        <input type="hidden" name="proxymethod" value="post">

                        <input type="hidden" name="product" value="nagiosxi">
                        <input type="hidden" name="version" value="5.2.3">
                        <input type="hidden" name="build" value="1449592399">

                        <label for="feedbackCommentBox">Comments:</label>
                        <textarea class="textarea form-control" name="comment" style="width: 100%; height: 100px;"></textarea>

                        <label for="feedbackNameBox">Your Name (Optional):</label>
                        <input type="text" size="30" name="name" id="feedbackNameBox" value="" class="textfield form-control">

                        <label for="feedbackEmailAddressBox">Your Email Address (Optional):</label>
                        <input type="text" size="30" name="email" id="feedbackEmailAddressBox" value="" class="textfield form-control">

                        <div>
                            <div class="fl" id="feedbackFormButtons">
                                <input type="submit" class="submitbutton btn btn-sm btn-primary" name="submitButton" value="Submit" id="submitFeedbackButton">
                            </div>
                            <div class="fr feedback-pp">
                                <a href="https://www.nagios.com/legal/privacypolicy/" target="_blank">Privacy Policy</a>
                            </div>
                            <div class="clear"></div>
                        </div>

                    </form>
                </div>
            </div>
        </div>
    </div>
<div id="popup_layer">
    <div id="popup_content">
        <div id="popup_close">
            <a id="close_popup_link" style="display: inline-block;" title="Close"><i class="fa fa-times" style="font-size: 16px;"></i></a>
        </div>
        <div id="popup_container">
        </div>
    </div>
</div>

        <div id="throbber" class="sk-spinner sk-spinner-center sk-spinner-three-bounce">
            <div class="sk-bounce1"></div>
            <div class="sk-bounce2"></div>
            <div class="sk-bounce3"></div>
        </div>
    </div>

            <div id="mainframe">

        
        <script>
// frame buster
if (top.location != self.location) {
    top.location = self.location.href;
}
</script>
<div class="login-table">
    <div class="login-cell left">

        <script>
        $(document).ready(function() {
            if ("" == '') {
                $('input[name="username"]').focus();
            } else {
                $('input[name="password"]').focus();
            }
        });
        </script>

        
        <div class="login-box">
            <div class="well">

                <form id="loginForm" method="post" action="/nagiosxi/login.php">
                    <input type='hidden' name='nsp' value='56a8f02b80cdf0ca30d8e8d321345533'>
                    <input type="hidden" name="page" value="auth">
                    <input type="hidden" name="debug" value="">
                    <input type="hidden" name="pageopt" value="login">
                    
                    <h1>Login</h1>

                    
                    <div style="margin-bottom: 10px;">
                        <input type="text" class="form-control" placeholder="Username" value="" name="username" id="usernameBox">
                    </div>
                    <div>
                        <input type="password" class="form-control" placeholder="Password" name="password" id="passwordBox">
                    </div>
                    <div>
                        <button type="submit" class="submitbutton login btn btn-sm btn-primary" id="loginButton" name="loginButton">Login</button>
                    </div>
                    
                    
                    
                </form>
                <p class="forgot-pw"><a href="?forgotpass">Forgot your password?</a></p>
            </div>
            <div class="select-language">
                <label>Select Language:</label>
                <div id='languageopts'>
                    <a href='?locale=en_US' class='locale-icon en_US' title='English'></a>
                    <a href='?locale=de_DE' class='locale-icon de_DE' title='German'></a>
                    <a href='?locale=es_ES' class='locale-icon es_ES' title='Spanish'></a>
                    <a href='?locale=fr_FR' class='locale-icon fr_FR' title='French'></a>
                    <a href='?locale=it_IT' class='locale-icon it_IT' title='Italian'></a>
                    <a href='?locale=ko_KR' class='locale-icon ko_KR' title='Korean'></a>
                    <a href='?locale=pt_PT' class='locale-icon pt_PT' title='Portuguese'></a>
                    <a href='?locale=ru_RU' class='locale-icon ru_RU' title='Russian'></a>
                    <a href='?locale=zh_CN' class='locale-icon zh_CN' title='Simplified Chinese'></a>
                    <a href='?locale=zh_TW' class='locale-icon zh_TW' title='Traditional Chinese'></a>
                    <a href='?locale=ja_JP' class='locale-icon ja_JP' title='Japanese'></a>
                    <a href='?locale=pl_PL' class='locale-icon pl_PL' title='Polish'></a>
                </div>
            </div>
        </div>
        <div class="clear"></div>

    </div>

    <div class="login-cell right">

        
        <img src="https://10.1.1.109/nagiosxi/images/loginsplash.png">
        <h3>About Nagios XI</h3>
        <p>
            Nagios XI is an enterprise-class monitoring and alerting solution that provides organizations with extended insight of their IT infrastructure before problems affect critical business processes.  For more information on Nagios XI, visit the <a href="//www.nagios.com/products/nagiosxi/" target="_blank">Nagios XI product page</a>.
        </p>
        <h3>Nagios Learning Opportunities</h3>
        <p>
            Learn about Nagios            <a href="//www.nagios.com/services/training"
               target="_blank"><strong>training</strong></a>
            and <a href="//www.nagios.com/services/certification" target="_blank">
                <strong>certification</strong></a>.
        </p>
        <p>
            Want to learn about how other experts are utilizing Nagios?  Don't miss your chance to attend the next            <a href="//go.nagios.com/nwcna" target="_blank"><strong>Nagios World Conference</strong></a>.
        </p>
        <h3>Contact Us</h3>
        <p>
            Have a question or technical problem? Contact us today:        </p>
        <table class="table table-condensed table-no-border" style="width: auto;">
            <tr>
                <td>Support:</td>
                <td><a href="//support.nagios.com/forum/" target="_blank">Online Support Forum</a></td>
            </tr>
            <tr>
                <td style="vertical-align: top;">Sales:</td>
                <td>
                    Phone: (651) 204-9102
                    <br>Fax: (651) 204-9103
                    <br>Email: sales@nagios.com
                </td>
            </tr>
            <tr>
                <td valign="top">Web:</td>
                <td><a href="//www.nagios.com/" target="_blank">www.nagios.com</a></td>
            </tr>
        </table>

        
    </div>
    <div class="clear"></div>
</div>

    </div><!--mainframe-->

    <!-- FOOTER START -->
<div id="footer">

    
    <div class="container-fluid">
        <div class="row">
            <div class="col-sm-6 footer-left">
                <a href="http://nagios.com/products/nagiosxi" target="new"><strong>Nagios XI</strong></a>                             </div>
            <div class="col-sm-6 footer-right">
                
                <a href="https://10.1.1.109/nagiosxi/about/">About</a> &nbsp;&nbsp;|&nbsp;&nbsp;
                <a href="https://10.1.1.109/nagiosxi/about/?legal">Legal</a> &nbsp;&nbsp;|&nbsp;&nbsp;
                Copyright &copy; 2008-2021 <a href="https://www.nagios.com/" target="_blank">Nagios Enterprises, LLC</a>
            </div>
        </div>
    </div>

    
    <script type="text/javascript">

        function get_tray_alert_content() {
            var optsarr = {
                "func": "get_tray_alert_html",
                "args": ""
            }
            var opts = array2json(optsarr);
            get_ajax_data_with_callback("getxicoreajax", opts, "process_tray_alert_content");
        }

        function process_tray_alert_content(edata) {
            data = unescape(edata);
            $("#tray_alerter_popup_content").html(data);

            var status = $("#tray_alerter_status").html();
            $("#tray_alerter").html(status);
        }

        $(document).ready(function () {

            get_tray_alert_content();

            $("#tray_alerter").everyTime(30000, "timer-tray_alerter", function (i) {
                get_tray_alert_content();
            });

            $("#tray_alerter").click(function () {
                var vis = $("#tray_alerter_popup").css("visibility");
                if (vis == "hidden") {
                    $("#tray_alerter_popup").css("visibility", "visible");
                } else {
                    $("#tray_alerter_popup").css("visibility", "hidden");
                }
            });

        });
    </script>

    <div id="keepalive"></div>

</div> <!-- end footer div -->

<!-- FOOTER END -->
    </div><!--page-->

    <noframes>
        <!-- This page requires a web browser which supports frames. -->
        <h2>Nagios XI</h2>
        <p align="center">
            <a href="https://www.nagios.com/">www.nagios.com</a><br>
            Copyright (c) 2009-2021 Nagios Enterprises, LLC<br>
        </p>
        <p>
            <i>Note: These pages require a browser which supports frames</i>
        </p>
    </noframes>

    
    <script type='text/javascript'>

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-2887186-1']);
  _gaq.push(['_setAllowLinker', true]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })(); 

</script>
    </body>
</html>


[-] The target is not exploitable. Target is not a Nagios XI application
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```
</p></details>